### ISSUE TITLE
Constrain the scaling factors of the two ranges

### ISSUE THREAD

**Issue by Shape-Kim** [hahnyuan/PTQ4ViT#8]
First of all, thank you for the great work and the official code.

I have one question. 

Where is the code implementation for constraining the scaling factors for post-softmax and post-gelu, i.e., ∆R2 = 2m∆R1, where m is an unsigned integer, in order for an efficient process.

I really appreciate for providing the code once again.

**Comment by SuperVan-Young** [comment]
PTQ4ViT adopts two classes `SoSMatMul` and `PostGeluLinear` in `quant_layers/`. `SoSMatMul` adopts a variable `split` to quantize post-softmax values. It indicates the split point of two ranges [0, split] and [split, 1], and is initialized as 2**m (m < 0). `PostGeluLinear` adopts an additional scaling factor `a_neg_interval` for the negative range, and the positive range's candidate intervals are initailized to satisfy the constraint.

**Comment by Shape-Kim** [comment]
Thank you for your quick response!

For post-softmax, I can find that split points satisfy the constraints throughout [the code](https://github.com/hahnyuan/PTQ4ViT/blob/main/quant_layers/matmul.py#:~:text=._initialize_intervals()-,A_split_candidates%20%3D%20torch.tensor(%5B2**(%2Di)%20for%20i%20in%20range(20)%5D).cuda(),-B_interval_candidates%20%3D%20torch).

However, I am still curious about the case of post-gelu. As I understand, a_interval might become **(2^m) * a_neg_interval**.

In the `PostGeluPTQSLQuantLinear`, I can find [the code](https://github.com/hahnyuan/PTQ4ViT/blob/main/quant_layers/linear.py#:~:text=input_interval_candidates%20%3D%20%20torch.tensor(%5Bself.eq_alpha%20%2B%20i*(self.eq_beta%20%2D%20self.eq_alpha)/self.eq_n%20for%20i%20in%20range(self.eq_n%20%2B%201)%5D).cuda().view(1%2C1%2C%2D1)%20*%20self.a_interval%5B0%5D.unsqueeze(%2D1)%20%23%20shape%3A%20n_a%2C1%2Ceq_n) for initializing the candidates for input interval. However, this does not satisfy the constraint. 

In order to check, I run the code for quantizing the ViT-S in 6/6-bit setting. For the first gelu layer, self.a_interval and self.a_neg_interval are 0.0053116 and 0.0289, respectively. Therefore, self.a_interval / self.a_neg_interval is 34.6432 which is not satisfying the constraint.

I am thinking that I misunderstand something about the code. Please answer my question. 
 
Thank you for your quick answer again.

**Comment by SuperVan-Young** [comment]
A_interval_candidates should be initialized with 2 ** m * initial a_neg_interval. Looks like a mistake pop up when merging the released version code, original code for experiment is kind of messy. 
Sorry for that if it bothers you. You can correct this if you find that troublesome. 
I might not have the chance for validating the results recently, since the server with the original experiment setup is constantly held up by other people these days, and I’m catching up another conference’s deadline. 
Really sorry for that once again, and thank you for pointing out. I’ll manage hotfix later.

**Comment by Shape-Kim** [comment]
Thank you for the reply again! 

I hope this problem can be fixed, and I'll cross my fingers for your current work!

**Comment by haiduo** [comment]
> Thank you for the reply again!
> 
> I hope this problem can be fixed, and I'll cross my fingers for your current work!

The implementation here seems to be hard-coded? Do you have a solution?
Please, Have you verified the modified results? Can you provide reference results? Thank you！

**Comment by haiduo** [comment]
[        self.a_interval.append(0.16997124254703522/self.a_qmax)](https://github.com/hahnyuan/PTQ4ViT/blob/b42c790c2811e7835d435a9f9a6c64a079a7483a/quant_layers/linear.py#L320)

Hi @SuperVan-Young , the scale of the negative value area to this? Is it the result of parameter adjustment or random?
